naga/src/lib.rs

```diff
@@ -2003,6 +2005,54 @@ pub enum Statement {
         /// [`SHADER_INT64_ATOMIC_ALL_OPS`]: crate::valid::Capabilities::SHADER_INT64_ATOMIC_ALL_OPS
         result: Option<Handle<Expression>>,
     },
+    /// Performs an atomic operation on a texel value of an image.
+    ///
+    /// Doing atomics on images with mipmaps is not supported, so there is no
+    /// `level` operand.
+    ///
+    /// This statement is a barrier for any operations on the corresponding
+    /// [`Expression::GlobalVariable`] for this image.
+    ImageAtomic {
+        /// The image to perform an atomic operation on. This must have type
+        /// [`Image`]. (This will necessarily be a [`GlobalVariable`] or
+        /// [`FunctionArgument`] expression, since no other expressions are
+        /// allowed to have that type.)
+        ///
+        /// [`Image`]: TypeInner::Image
+        /// [`GlobalVariable`]: Expression::GlobalVariable
+        /// [`FunctionArgument`]: Expression::FunctionArgument
+        image: Handle<Expression>,
+
+        /// The coordinate of the texel we wish to load. This must be a scalar
+        /// for [`D1`] images, a [`Bi`] vector for [`D2`] images, and a [`Tri`]
+        /// vector for [`D3`] images. (sample indices are supplied separately.)
+        /// Its component type must be [`Sint`].
+        ///
+        /// If this image is arrayed, [`D1`] images require a [`Bi`] vector and
+        /// [`D2`] images require a [`Tri`] vector.
+        ///
```

Comment:
Including the array index in the coordinates makes `ImageAtomic` unlike `Statement::ImageStore` and `Expression::ImageLoad`, where the convention is to treat the array index as a distinct parameter. This should be changed to be like the others.

In the SPIR-V backend, you can use `BlockContext::write_image_coordinates` to produce the combined "Coordinate" value that `OpImageTexelPointer` wants.

Reply:
this was probably refactored after i copy pasted from those, as i did base it off of those